### PR TITLE
chore(remap transform): Add non-interpolated VRL error message to dropped event

### PR DESCRIPTION
**Pull request by 001wwang**
Part of #10686 

This PR adds a non-interpolated error message from VRL as a `raw_message` field in the dropped event's `metadata`. This allows access to the original user-provided message for `assert` and `assert_eq`.

### CODE REVIEW

File: src/transforms/remap.rs [hunk 1]
```diff
@@ -132,12 +136,18 @@ impl Remap {
     fn annotate_dropped(&self, event: &mut Event, reason: &str, error: ExpressionError) {
         match event {
             Event::Log(ref mut log) => {
+                let raw_message = error
+                    .labels()
+                    .pop()
+                    .expect("at least one label from error")
```
Comment (spencergilbert) [inline src/transforms/remap.rs:142]: What labels does an `error` have in this case?

Comment (001wwang) [reply src/transforms/remap.rs:142]: From my understanding, `labels` is populated with successive error messages from a function call and its potentially nested function calls: https://github.com/vectordotdev/vector/blob/master/lib/vrl/compiler/src/expression/function_call.rs#L228. With `pop()` we get the top most error message prior to it being interpolated into `"function call error..."`, but if there were nested calls that errored, even `raw_message` might contain `"function call error..."`.

Comment (spencergilbert) [reply src/transforms/remap.rs:142]: IMHO an odd name for what it contains, thanks!

**Review by lukesteensen — COMMENTED** [review]
This implementation seems reasonable to me, I'm just a little unsure about the naming. Having `message` and `raw_message` fields seems a little confusing to me, but I don't have a great idea for how better to differentiate these.

**Comment by 001wwang** [comment]
> This implementation seems reasonable to me, I'm just a little unsure about the naming. Having `message` and `raw_message` fields seems a little confusing to me, but I don't have a great idea for how better to differentiate these.

Agreed, the naming isn't ideal. Is there a reason we want both the original and interpolated messages? Could we get away with just setting `message` to be the non-interpolated message?

 Like you'd mentioned previously, we could have a nested structure that makes it a bit clearer that the fields are related. Though, it doesn't solve the naming issue.
```
{
  message: {
    interpolated: "function call error... : custom error"
    original: "custom error"
  }
}
```

**Comment by spencergilbert** [comment]
> > This implementation seems reasonable to me, I'm just a little unsure about the naming. Having `message` and `raw_message` fields seems a little confusing to me, but I don't have a great idea for how better to differentiate these.
> 
> Agreed, the naming isn't ideal. Is there a reason we want both the original and interpolated messages? Could we get away with just setting `message` to be the non-interpolated message?
> 
> Like you'd mentioned previously, we could have a nested structure that makes it a bit clearer that the fields are related. Though, it doesn't solve the naming issue.
> 
> ```
> {
>   message: {
>     interpolated: "function call error... : custom error"
>     original: "custom error"
>   }
> }
> ```

I'd suggest avoiding changing the `type` of `.message` - while not a hard rule it'd generally be assumed to be a `string`

**Comment by lukesteensen** [comment]
> Is there a reason we want both the original and interpolated messages? Could we get away with just setting `message` to be the non-interpolated message?

That seems reasonable to me. In cases where the user provides a message, just use that. Otherwise, use the more interpolated one.

**Review by jszwedko — COMMENTED** [review]
Nice! I like the idea to use the user message, if provided. If there wasn't a provided message though (ala `assert_eq`) I think we may still want the completely wrapped message.

That is we'd want:

```
function call error for \"string\" at (160:175): expected \"string\", got \"integer\"
```

Rather than:

```
expected \"string\", got \"integer\"
```

I'm not 100% sure how easy it will be to determine the difference for errors though.

Failing that, I prefer the original implementation that returned the wrapped errors even though it results in the user's assert message being wrapped in a `function call error`.

**Comment by 001wwang** [comment]
> I'm not 100% sure how easy it will be to determine the difference for errors though: which are from `assert_eq` and `assert` vs. runtime errors.
> 
> Failing that, I prefer the original implementation that returned the wrapped errors even though it results in the user's assert message being wrapped in a `function call error`.

Gotcha, I realize with this implementation using `error.labels()`, it's impossible to determine whether the unwrapped error is a user message or a generated message. The current behavior will always give you the unwrapped error (unless `labels` is empty, which I think is not possible). Deeper VRL changes could enable it (modify `ExpressionError::Error.message` to use, say, a `Message` type that we can bake more info into instead of just a `String`, maybe change `Label` as well) but not sure if that'd have broader consequences. I can experiment with this.

Otherwise, I can roll this back to the iteration that returns both a wrapped error and unwrapped error in the event `metadata` if that's better for a first pass.

**Review by jszwedko — APPROVED** [review]
I'm content with leaving in the `function call error` part of the message in this first pass. In the future we can revisit this. I think it is somewhat an artifact of how `assert(_eq)!` works in that it raises an error if the assertion is false.

**Comment by 001wwang** [comment]
In fiddling around with `ErrorExpression::Error`'s message field, I realized adding to the existing `Note` enum might be more straightforward and require fewer changes. The latest iteration in feaec5a gets us to the desired the UX of
- If a user provides a message to `assert`/`assert_eq`, we use it for the `message` field
- Otherwise, we use the standard error for the `message` field

The only worry I have is if this abuses what `Note` is intended for (more diagnostic purposes?) but otherwise, I'm happy with this.

**Review by jszwedko — APPROVED** [review]
Nice! I'm glad you were able to find a way to use just the custom message if set.

**Pull request merged**